### ISSUE TITLE
Race condition (?) with image loading

### ISSUE THREAD

**Issue by nachtjasmin** [mastodon/mastodon#23800]
### Steps to reproduce the problem

1. Open https://queer.group/@twitchbunny@pettingzoo.co/109905500238458878.
2. Wait, until the page is loaded.
3. The image appears and then disappears. Reproducible on our instance.

### Expected behaviour

The image should be displayed.

### Actual behaviour

It was displayed and got replaced with the "not available" screen.

### Detailed description

When I am opening the above link, the image is loaded first and later replaced with the "not available" note. However, the image is there. It exists on the server. It's even fetched as seen in the developer tools of Firefox.

But it's aborted by the browser. I don't know whether this is a race condition, but from what I've researched about `NS_BINDING_ABORTED`, it sounds like one.

![Screenshot of three network requests. Two are successful, the third not due to NS_BINDING_ABORTED.](https://user-images.githubusercontent.com/98263758/220577829-bad81db4-f44a-440c-ad02-4c1c3d59ca38.png)

[mastodon-weird-behaviour.webm](https://user-images.githubusercontent.com/98263758/220581371-cc27f7ad-afa4-4e9c-8d90-ce3d397c1eac.webm)

### Specifications

Software | Version
-- | --
Mastodon/Hometown | 4.0.2+hometown-1.1.1
Ruby | 3.0.4p208
PostgreSQL | 15.2
Redis | 6.0.16
Browser | Firefox 110 on Arch Linux

And yes, I know it *could* be an issue in the fork, however I don't think that this is the case here. None of the features of the fork are involved in this issue.

**Comment by chloe-the-catgirl** [comment]
If you want, I can create a temporary account for you on the instance.

I'm happy to support you as much as I can, because I don't have the expertise to look for errors in the code myself. 🙈

**Comment by vmstan** [comment]
@nachtjasmin is this still an issue?

**Comment by nachtjasmin** [comment]
Good question, haven't encountered it anymore. I think, it's okay to close this issue.

**Comment by jdeluyck** [comment]
I can confirm this issue on my own instance.

**Comment by vmstan** [comment]
Mastodon 4.2.1?

**Comment by jdeluyck** [comment]
> Mastodon 4.2.1?

Yes. Instance is on https://fedi.kcore.org, I can reproduce it on both Firefox linux and Firefox Android (the two platforms I use daily).

I can create an account for someone here to test if need be.

**Comment by ClearlyClaire** [comment]
> Yes. Instance is on https://fedi.kcore.org, I can reproduce it on both Firefox linux and Firefox Android (the two platforms I use daily).

I cannot find any public-facing media to try to see how it works.

Are you using file system storage, or external object storage? Do you have a CDN in front?

> I can create an account for someone here to test if need be.

That could be useful indeed!

**Comment by jdeluyck** [comment]
> Are you using file system storage, or external object storage? Do you have a CDN in front?

No CDN. All storage is local.

> > I can create an account for someone here to test if need be.
> 
> That could be useful indeed!

If I can get an email address somewhere for the invite, that would be great ;) so I know where to send the invite link to.

**Comment by ClearlyClaire** [comment]
> If I can get an email address somewhere for the invite, that would be great ;) so I know where to send the invite link to.

Mail to `claire@joinmastodon.org`

**Comment by jdeluyck** [comment]
> > If I can get an email address somewhere for the invite, that would be great ;) so I know where to send the invite link to.
> 
> Mail to `claire@joinmastodon.org`

It's sent.

Is there some log I can peek at meanwhile that would help finding what goes wrong when I can trigger it?
It seems to be very reproducible when I go to a random person's profile (on my instance), go to media, and start clicking on images. I see the progress bar while it's being downloaded to my instance, than a short flash while the image is visible, and then it dissapears.

I've now been able to reproduce this on 2 profiles.

**Comment by ClearlyClaire** [comment]
I did receive your invite mail, but I am not getting a confirmation email from your server.

**Comment by jdeluyck** [comment]
I've resend a link as the previous one expired.

**Comment by ClearlyClaire** [comment]
It hadn't expired, I was able to create an account, but not to confirm it, because I have not received the confirmation email.

**Comment by jdeluyck** [comment]
I've activated the accounts. Will check the mail stuff later, as I thought this did work.

**Comment by ClearlyClaire** [comment]
Alright, I'm signed in, and can post media attachments, browse the feeds and view media attachments without issue. Is there something specific that would bring the issue up?

**Comment by jdeluyck** [comment]
How I can most easily reproduce it:

Pick any profile page of $randomperson, go to their media, click on the images. I just tried it on someone at random from my timeline and at the second picture I'm already hitting this issue.

**Comment by ClearlyClaire** [comment]
Ah, I think what happens is you are running into rate limits when re-fetching old remote media evicted from your local cache. Are you getting 429 errors per any chance?

**Comment by jdeluyck** [comment]
Hah. I just went looking in the logs of mastodon-web.service, and everytime I encounter this I get

```
Nov 07 17:58:55 fedi bundle[1427408]: I, [2023-11-07T17:58:55.999728 #1427408]  INFO -- : [df28c8f7-ef55-4aac-be63-9c75c130ef3c] Rate limit hit (throttle): 78.22.67.165 GET /media_proxy/111324341922382606/small
Nov 07 17:58:56 fedi bundle[1427400]: I, [2023-11-07T17:58:56.005725 #1427400]  INFO -- : [7717ea0c-2bb2-4574-985d-2163ff2c6d42] Rate limit hit (throttle): <my-ip> GET /media_proxy/111324341922382606/original
Nov 07 17:58:56 fedi bundle[1427408]: I, [2023-11-07T17:58:56.085512 #1427408]  INFO -- : [822df504-3d34-4f4a-bcdc-a7e0f1c77b4d] Rate limit hit (throttle): <my-ip> GET /media_proxy/111324341922382606/original
```

Now, I waited a minute, and now it's no issue whatsoever.

**Comment by ClearlyClaire** [comment]
I assume you have an aggressively low “Media cache retention period” in `/admin/settings/content_retention`? Then the behavior is kinda expected, although it would make sense to have better error handling client-side, and reconsider the rate limiting implementation.

**Comment by nachtjasmin** [comment]
What is considered "aggressively low"? For our instance, it's set to 14 days right now.

**Comment by jdeluyck** [comment]
For my instance it's set at 7 days.

**Comment by smiba** [comment]
Would it maybe be a solution to increase media_proxy limits for authenticated users? Especially for instances with manual account activation (or even single-user instances) it would be beneficial to do so. 

https://github.com/mastodon/mastodon/blob/5fc4ae7c5f9667b335e63f97cacaa1efe5f7a6d5/config/initializers/rack_attack.rb#L85-L87
